File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        diameter = 0

        def max_diameter(root):
            nonlocal diameter

            if not root: return -1

            left_height = max_diameter(root.left) + 1
            right_height = max_diameter(root.right) + 1

            diameter = max(diameter, left_height + right_height)

            return max(left_height, right_height)

        max_diameter(root)
        return diameter

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        min_ = std::min(p->val, q->val);
        max_ = std::max(p->val, q->val);

        if (root->val >= min_ && root->val <= max_) return root;

        if (root->val > max_) return lowestCommonAncestor(root->left, p, q);
        else if (root->val < min_) return lowestCommonAncestor(root->right, p, q);
        

    }
    
private:
    int min_ = std::numeric_limits<int>::max();
    int max_ = std::numeric_limits<int>::min();
};
